[Draft] New Block Reconstruction Protocol Proposal

walldiss opened the pull request:
This PR introduces a draft proposal for a new block reconstruction protocol aimed at improving network efficiency and scalability.

Key improvements:
- Reduces duplicate requests and bandwidth usage
- Structured bitmap sharing using Roaring Bitmaps
- Clear separation of client/server responsibilities
- Robust failure handling with peer penalties
- Optimized proof packaging and sharing

The document serves as a starting point for discussion and will be later split into CIP and ADR.

TODO:
- [ ] Add testing plan
- [ ] Add false positive triggered reconstruction impact scenario. Potential harm and mitigation

walldiss committed:
reconstruction draft

walldiss committed:
reconstruction draft

musalbas commented:
Looks generally fine to me directionally. Networking protocol should probably be in a CIP as it's relevant to other clients like lumina etc.

Wondertan commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,286 @@
+### Core Protocol Components
+
+#### 1. Connection Management
+
+Handshakes, Node Type Identification
+- Node type must be declared during connection handshake by each of connection nodes
+- Active connection state maintenance on full nodes. List should be maintained by node type
+- Connection state should allow subscription for updated
```
Whats the rationale behind sharing your node type? Is it possible to avoid sharing this information with peers?

There is a privacy concern we discussed a while ago with @musalbas and @adlerjohn. FNs ideally should be indistinguishable from LNs on p2p network, so they validators don't detect them.

If we don't think that's important anymore or sharing type is unavoidable, I also propose to cut the handshake roundtrip and rely on libp2p toolset for that.

musalbas replied:
That is indeed a goal, but the problem here is that it's implicit in this protocol that only full nodes request bitmaps from others, regardless of how you identify in the handshake.

FNs pretending to be LNs (to reconstruct) would only work if the FNs spam sample requests to the block producer, which isn't something implemented yet afaik. (See https://forum.celestia.org/t/security-levels-for-data-availability-for-light-nodes/919#level-3-data-availability-sampling-without-an-honest-minority-of-light-nodes-6)

Wondertan replied:
> if the FNs spam sample requests to the block producer, which isn't something implemented yet afaik. 

Do you think this should be in scope of current iteration of reconstruction protocol?

musalbas replied:
I think not in this iteration, because it seems we're doing pretty good on increasing number of light nodes, so we could just focus directly on having a robust construction protocol and achieving Level 4 light nodes. It might be nice to have later as an option, though, but not high priority right now.

walldiss replied:
Good point, handshake is not necessary for initial basic implementation. It will be needed in peer scoring optimisation. When FN would need to know the type of peer for prioritisation of requests. I'll mark handshake protocol as future optimisation.

renaynay commented on docs/adr/reconstruction.md:
````diff
@@ -0,0 +1,286 @@
+```go
+type RemoteState struct {
+    coords [][]peers    // Peer lists by coordinates
+    rows []peers        // Peers with row data
+    cols []peers        // Peers with column data
````
how is this actually going to be structured? 

you keep a []peer.ID per which row they claim to have? So index 0 []peer.ID will be peers that claim to have row 0?

walldiss replied:
Yes. The idea is to have efficient look up for peers who have full row or full col.

renaynay commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,286 @@
+
+Local state:
+- InProgress() -> bitmap. Tracks ongoing fetch sessions to prevent requesting duplicates
+- Have() -> bitmap.  Tracks 
```
```suggestion
- Have() -> bitmap.  Tracks node's locally available samples
```

renaynay commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,286 @@
+    - Pre-confirmation threshold
+    - Combination of conditions
+
+4.1.1 what and were (to request) biggest question.
```
what and where ?

renaynay commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,286 @@
+4.2.1. Request 
+
+- Update InProgress bitmap before request initiation
+- Use shrex for data retrieval. Send bitmap for data request. Shrex has built-in support for requesting rows and samplies, however bitmap-based data retrieval is more general and would be easier to support in future, because it allows requesting multiples pieces of data with single request instead of multiple requests
```
```suggestion
- Use shrex for data retrieval. Send bitmap for data request. Shrex has built-in support for requesting rows and samples, however bitmap-based data retrieval is more general and would be easier to support in future, because it allows requesting multiples pieces of data with single request instead of multiple requests
```

renaynay commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,286 @@
+Optiomisations:
+- Server may split response into multiple parts
+Each part contains packed samples in Range format.
+- Each part of shoold have specified message prefix. It would allow client to identify sample container type, which would help to maintain backwards compatibility in case of breaking changes in packing algorithm
```
```suggestion
- Each part of should have specified message prefix. It would allow client to identify sample container type, which would help to maintain backwards compatibility in case of breaking changes in packing algorithm
```

renaynay commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,286 @@
+4.2.2. Response 
+- (base version) Server should respond with samples with proofs
+
+Optiomisations:
```
```suggestion
Optimisations:
```

renaynay commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,286 @@
+### Server-Side Implementation
+
+#### 1. Storage Interface
+New storage format needs to be implemented for efficient storage of sampels proofs. The format will be used
```
```suggestion
New storage format needs to be implemented for efficient storage of sample proofs. The format will be used
```

renaynay commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,286 @@
+
+#### 1. Storage Interface
+New storage format needs to be implemented for efficient storage of sampels proofs. The format will be used
+initially used for storing ongoing reconstruction process and later can be used for light node storage.
```
```suggestion
initially for storing ongoing reconstruction process and later can be used for light node storage.
```

renaynay reviewed (commented):
Would this proposal also require some modifications in discovery + peer management as it currently stands? If so, can you please outline what that would look like?

walldiss committed:
refactor

walldiss committed:
refactor 2

walldiss commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,317 @@
+3. Fallback mechanisms
+
+
+
```
Add testing plan / requirements

walldiss reviewed (commented):
Feedback notes from sync call

walldiss committed:
added:
- return handshake based on user agent.
- Do not subscribe for samples from LN
- Request from LN in batches

walldiss committed:
added:
- return handshake based on user agent.
- Do not subscribe for samples from LN
- Request from LN in batches

walldiss committed:
Add reconstruction step by step diagram

walldiss committed:
formatting improvements

walldiss committed:
add sampling protocol motivation

walldiss committed:
add false positive

musalbas commented on docs/adr/reconstruction.md:
```diff
@@ -0,0 +1,488 @@
+
+### No Bitmap Subscription From Light Nodes: Why?
+
+Subscribing to bitmaps from Light Nodes would allow more precise deduplication of requested samples, but it also introduces additional complexity and round-trip overhead. Since each Light Node holds relatively few samples, the probability of overlap is low. Instead, the Full Node can send an inverse “have” bitmap in its request to indicate which samples are still needed.
```
Does this mean that the FN needs to regularly send the bitmap (or a diff of the bitmap) again to LNs as the FN collects more samples? Or is it not necessary as most LNs only have a few samples anyway?

walldiss replied:
Good question! The short answer is that **updates are not needed.**

Because each Light Node typically holds only a few samples, a single request round-trip to the LN is usually sufficient to retrieve everything it has. If the LN sampled only some of 16 coordinates it can provide none. So the FN can retry later with an increasing backoff. This approach allows the FN to determine whether a particular LN has fully sampled the block and no additional bitmap exchanges are required. Essentially, once the LN reports having all samples (or delivers all it has), the FN can safely move on.

musalbas replied:
How does the FN know if the LN has all the samples though? SamplesRequest struct doesn't provide for this.

musalbas replied:
I think this will result cases where the block cannot be reconstructed even if LNs have enough shares to do so -- for example if 1/4 of the block is withheld, most LN sampling will fail, til reconstruction happens. It's a chicken and egg situation...

walldiss replied:
Yes, that’s correct: if 1/4 of the block is withheld, LNs can’t finish sampling and therefore won’t serve any shares. 99% of LNs will treat the block as unavailable. Although it may seem like a “chicken and egg” issue, the protocol is actually protecting LNs from finalizing incomplete blocks. The block could be recovered if the missing shares eventually surface, but until then, it remains withheld. The assumption here is that protocol’s main goal is to prevent LNs from accepting blocks that aren’t fully available.

Do you think this scenario is a viable attack vector? In given case random 1% of LN will get into thinking that block is available, which is the same as the chance of false positive sampling

musalbas replied:
My point is that in such a scenario, LNs should be sharing their partial samples with FNs, in case FNs can reconstruct the missing parts of the block, and complete the rest of their samples.
